assignment2.rb

```diff
@@ -0,0 +1,192 @@
+#Blocks, Procks and Lambda
+
+#All three mechanisms allow you to encapsulate code for reuse, but they have subtle differences in their behavior and how they handle certain situations. 
+#Blocks are used for inline code within method invocations, 
+#procs are general-purpose saved blocks,
+# and lambdas are special type of procs with stricter argument handling.
+
+#Use Blocks, Procs and Lambda each to write a program which takes array of integers as an input and finds the count of even numbers in the array
+
+
+#using Blocks
+def count_even_numbers(array)
+    count = 0
+    array.each do |num|
+      count += 1 if num.even?
+    end
+    count
+  end
+  
+ numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+  puts "Count of even numbers: #{count_even_numbers(numbers)}"
+
+  #using Procs
+  count_even = Proc.new { |num| num.even? }
+def count_numbers(array, condition)
+  array.count(&condition)
+end
+
+
+numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+puts "Count of even numbers: #{count_numbers(numbers, count_even)}"
+
+#using lambda
+
+count_even = lambda { |num| num.even? }
+def count_numbers(array, condition)
+  array.count(&condition)
+end
+
+
+numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+puts "Count of even numbers: #{count_numbers(numbers, count_even)}"
+
+
+#####################################################################################################################################################
+
+require 'csv'
+
+class Employee
+  attr_accessor :name, :age, :department, :division, :designation
+
+  def initialize(name, age, department, division, designation)
+    @name = name
+    @age = age
+    @department = department
+    @division = division
+    @designation = designation
+  end
+end
+
+class Department
+  attr_accessor :name, :department_head, :employees
+
+  def initialize(name, department_head)
+    @name = name
+    @department_head = department_head
+    @employees = []
+  end
+end
+
+class Project
+  attr_accessor :name, :type, :clients, :profit
+
+  def initialize(name, type, clients, profit)
+    @name = name
+    @type = type
+    @clients = clients
+    @profit = profit
+  end
+
+end
+
+class Company
+  attr_accessor :departments, :projects, :yearly_revenue, :service_profit, :product_profit
+
+  def initialize
+    @departments = []
+    @projects = []
+    @yearly_revenue = 0
+    @service_profit = 0
+    @product_profit = 0
+  end
+
+  def add_department(department)
+    @departments << department
+  end
+
+  def add_project(project)
+    @projects << project
+  end
+
+  def add_employee(hr, employee)
```

Comment:
Company class shouldn't add employees. Please remove this method.

Reply:
Removed